From 078ae4e517813abd9976dbb1d4984cb9d52de1ce Mon Sep 17 00:00:00 2001
From: t <tristan75015@gmail.com>
Date: Wed, 14 Oct 2026 20:01:32 +0200
Subject: [PATCH] Initialize functions and post-processings: pass the behaviour
 options

The options of the behaviour (stress measure and tangent operator of
finite strain behaviours) are now copied after the first value of the
K array before calling an initialize function or a post-processing, as
integrate does. TFEL reads them in these functions.

The point by point functions of the material data managers did not set
the K array at all: finite strain behaviours dereferenced a null
pointer. A local array is used.
---
 include/MGIS/Behaviour/Integrate.hxx |  9 ++++++
 src/Integrate.cxx                    | 42 ++++++++++++++++++++--------
 2 files changed, 40 insertions(+), 11 deletions(-)

diff --git a/include/MGIS/Behaviour/Integrate.hxx b/include/MGIS/Behaviour/Integrate.hxx
index 653f936..a2c20d3 100644
--- a/include/MGIS/Behaviour/Integrate.hxx
+++ b/include/MGIS/Behaviour/Integrate.hxx
@@ -177,6 +177,9 @@ namespace mgis::behaviour {
    * \param[in] b: behaviour
    * \param[in] n: name of the initialize function
    * \param[in] inputs: inputs of the initialize function
+   * \note as by `integrate`, the options of the behaviour (see
+   * `Behaviour::options`) are copied in the `K` array of the behaviour data
+   * view, after its first value.
    * \note Due to the structure of the `BehaviourDataView` structure in which
    * the state at the beginning of the time step is immutable, only the state
    * variables at the end of the time step can be updated. Hence, for
@@ -196,6 +199,9 @@ namespace mgis::behaviour {
    * \param[in] b: behaviour
    * \param[in] n: name of the initialize function
    * \note the initialize function is expected to have no inputs
+   * \note as by `integrate`, the options of the behaviour (see
+   * `Behaviour::options`) are copied in the `K` array of the behaviour data
+   * view, after its first value.
    * \note Due to the structure of the `BehaviourDataView` structure in which
    * the state at the beginning of the time step is immutable, only the state
    * variables at the end of the time step can be updated. Hence, for
@@ -790,6 +796,9 @@ namespace mgis::behaviour {
    * \param[in,out] d: behaviour data
    * \param[in] b: behaviour
    * \param[in] n: name of the post-processing
+   * \note as by `integrate`, the options of the behaviour (see
+   * `Behaviour::options`) are copied in the `K` array of the behaviour data
+   * view, after its first value.
    * \note an exception is thrown for a behaviour generated by the
    * `generic-parallel` interface, which only provides functions of batches
    * of integration points.
diff --git a/src/Integrate.cxx b/src/Integrate.cxx
index 6ad44aa..5586726 100644
--- a/src/Integrate.cxx
+++ b/src/Integrate.cxx
@@ -1020,6 +1020,18 @@ namespace mgis::behaviour::internals {
     return v;
   }  // end of initializeBehaviourDataView
 
+  /*!
+   * \brief pass the options of the behaviour (stress measure and tangent
+   * operator of finite strain behaviours) to an initialize function or a
+   * post-processing, after the first value of the `K` array, as `integrate`
+   * does
+   * \param[in,out] d: behaviour data view
+   * \param[in] b: behaviour
+   */
+  static void setBehaviourOptions(BehaviourDataView& d, const Behaviour& b) {
+    std::copy(b.options.begin(), b.options.end(), d.K + 1);
+  }  // end of setBehaviourOptions
+
   /*!
    * \brief integrate the behaviour point by point over the points `[b, e)`
    * of a material data view.
@@ -1162,6 +1174,7 @@ namespace mgis::behaviour::internals {
   template <bool gather>
   static BehaviourIntegrationResult executeInitializeFunctionPointwise(
       BehaviourIntegrationWorkSpace& ws,
+      const Behaviour& behaviour,
       const PointwiseData& data,
       const BehaviourEvaluators& evaluators,
       const BehaviourInitializeFunction& f,
@@ -1170,6 +1183,9 @@ namespace mgis::behaviour::internals {
       const size_type e) {
     auto d = initializeBehaviourDataView(ws);
     d.rdt = nullptr;
+    real bopts[Behaviour::nopts + 1] = {};  // options passed to the behaviour
+    d.K = &bopts[0];
+    setBehaviourOptions(d, behaviour);
     auto r = BehaviourIntegrationResult{};
     for (auto i = b; i != e; ++i) {
       evaluators.evaluate(ws, i);
@@ -1227,11 +1243,11 @@ namespace mgis::behaviour::internals {
     const auto i =
         PointValues<const real>(inputs, ws.initialize_function_inputs);
     if (data.gathers || i.gathers()) {
-      return executeInitializeFunctionPointwise<true>(ws, data, evaluators, f,
-                                                      i, b, e);
+      return executeInitializeFunctionPointwise<true>(ws, *(v.b), data,
+                                                      evaluators, f, i, b, e);
     }
-    return executeInitializeFunctionPointwise<false>(ws, data, evaluators, f, i,
-                                                     b, e);
+    return executeInitializeFunctionPointwise<false>(ws, *(v.b), data,
+                                                     evaluators, f, i, b, e);
   }  // end of executeInitializeFunctionCheckedPointwise
 
   /*!
@@ -1258,6 +1274,7 @@ namespace mgis::behaviour::internals {
   static BehaviourIntegrationResult executePostProcessingPointwise(
       const PointValues<real>& outputs,
       BehaviourIntegrationWorkSpace& ws,
+      const Behaviour& behaviour,
       const PointwiseData& data,
       const BehaviourEvaluators& evaluators,
       const BehaviourPostProcessing& p,
@@ -1265,6 +1282,9 @@ namespace mgis::behaviour::internals {
       const size_type e) {
     auto d = initializeBehaviourDataView(ws);
     d.rdt = nullptr;
+    real bopts[Behaviour::nopts + 1] = {};  // options passed to the behaviour
+    d.K = &bopts[0];
+    setBehaviourOptions(d, behaviour);
     auto r = BehaviourIntegrationResult{};
     for (auto i = b; i != e; ++i) {
       evaluators.evaluate(ws, i);
@@ -1323,11 +1343,11 @@ namespace mgis::behaviour::internals {
     const auto data = PointwiseData(ws, v, false, false);
     const auto o = PointValues<real>(outputs, ws.postprocessing_outputs);
     if (data.gathers || o.gathers()) {
-      return executePostProcessingPointwise<true>(o, ws, data, evaluators, p, b,
-                                                  e);
+      return executePostProcessingPointwise<true>(o, ws, *(v.b), data,
+                                                  evaluators, p, b, e);
     }
-    return executePostProcessingPointwise<false>(o, ws, data, evaluators, p, b,
-                                                 e);
+    return executePostProcessingPointwise<false>(o, ws, *(v.b), data,
+                                                 evaluators, p, b, e);
   }  // end of executePostProcessingCheckedPointwise
 
   /*!
@@ -1449,7 +1469,7 @@ namespace mgis::behaviour {
     if (ifct.f == nullptr) {
       internals::raiseMissingFunctionOfAPoint("executeInitializeFunction");
     }
-    std::copy(b.options.begin(), b.options.end(), d.K + 1);
+    internals::setBehaviourOptions(d, b);
     const auto& f = ifct.f;
     return f(&d, nullptr);
   }  // end of executeInitializeFunction
@@ -1468,7 +1488,7 @@ namespace mgis::behaviour {
     if (ifct.f == nullptr) {
       internals::raiseMissingFunctionOfAPoint("executeInitializeFunction");
     }
-    std::copy(b.options.begin(), b.options.end(), d.K + 1);
+    internals::setBehaviourOptions(d, b);
     const auto& f = ifct.f;
     return f(&d, inputs.data());
   }  // end of executeInitializeFunction
@@ -2428,7 +2448,7 @@ namespace mgis::behaviour {
     if (p.f == nullptr) {
       internals::raiseMissingFunctionOfAPoint("executePostProcessing");
     }
-    std::copy(b.options.begin(), b.options.end(), d.K + 1);
+    internals::setBehaviourOptions(d, b);
     return (p.f)(outputs.data(), &d);
   }  // end of executePostProcessing
 
-- 
2.43.0

